src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs

```diff
@@ -317,26 +317,31 @@ private static MethodSymbol GetSynthesizedMethod(CommonPEModuleBuilder moduleBui
             }
             else
             {
+                formatSpecifiers = null;
+
                 // Try to parse as an expression. If that fails, parse as a statement.
-                var exprDiagnostics = DiagnosticBag.GetInstance();
+                var statementDiagnostics = DiagnosticBag.GetInstance();
+                var statementSyntax = expr.ParseStatement(statementDiagnostics);
+                Debug.Assert((statementSyntax == null) || !statementDiagnostics.HasAnyErrors());
+                statementDiagnostics.Free();
+                if (statementSyntax != null && !statementSyntax.IsKind(SyntaxKind.ExpressionStatement)) // Prefer to parse expression statements as expressions.
+                {
+                    if (statementSyntax.Kind() != SyntaxKind.LocalDeclarationStatement)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_ExpressionOrDeclarationExpected, Location.None);
+                        return null;
+                    }
+                    return statementSyntax;
+                }
+
                 ReadOnlyCollection<string> exprFormatSpecifiers;
-                CSharpSyntaxNode syntax = expr.ParseExpression(exprDiagnostics, allowFormatSpecifiers: true, formatSpecifiers: out exprFormatSpecifiers);
-                Debug.Assert((syntax == null) || !exprDiagnostics.HasAnyErrors());
-                exprDiagnostics.Free();
-                if (syntax != null)
+                var expressionSyntax = expr.ParseExpression(diagnostics, allowFormatSpecifiers: true, formatSpecifiers: out exprFormatSpecifiers);
+                if (expressionSyntax != null)
```

Comment:
Is the 'if' still needed or could this be "return expr.ParseExpression(..., out formatSpecifiers);"?